# Mini Project: Recommendation Engines

Recommendation engines are algorithms designed to provide personalized suggestions or recommendations to users. These systems analyze user behavior, preferences, and interactions with items (products, movies, music, articles, etc.) to predict and offer items that users are likely to be interested in. Recommendation engines play a crucial role in enhancing user experience, driving engagement, and increasing conversion rates in various applications, including e-commerce, entertainment, content platforms, and more.

There are generally two approaches taken in collaborative filtering and content-based recommendation engines:

**1. Collaborative Filtering:**
Collaborative Filtering is a popular approach to building recommendation systems that leverages the collective behavior of users to make personalized recommendations. It is based on the idea that users who have agreed in the past will likely agree in the future. There are two main types of collaborative filtering:

- **User-based Collaborative Filtering:** This method finds users similar to the target user based on their past interactions (e.g., ratings or purchases). It then recommends items that similar users have liked but the target user has not interacted with yet.

- **Item-based Collaborative Filtering:** In this approach, the system identifies similar items based on user interactions. It recommends items that are similar to the ones the target user has already liked or interacted with.

Collaborative filtering does not require any explicit information about items but relies on the similarity between users or items. It is effective in capturing complex patterns and can provide serendipitous recommendations. However, it suffers from the cold-start problem (i.e., difficulty in recommending to new users or items with no interactions) and scalability challenges in large datasets.

**2. Content-Based Recommendation:**
Content-based recommendation is an alternative approach to building recommendation systems that focuses on the attributes or features of items and users. It leverages the characteristics of items to make recommendations. The key steps involved in content-based recommendation are:

- **Feature Extraction:** For each item, relevant features are extracted. For movies, these features could be genre, director, actors, and plot summary.

- **User Profile:** A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.

- **Similarity Calculation:** The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity or Euclidean distance.

- **Recommendation:** Items that are most similar to the user profile are recommended to the user.

Content-based recommendation systems are less affected by the cold-start problem as they can still recommend items based on their features. They are also more interpretable as they rely on item attributes. However, they may miss out on providing serendipitous recommendations and can be limited by the quality of feature extraction and user profiles.

**Choosing Between Collaborative Filtering and Content-Based:**
Both collaborative filtering and content-based approaches have their strengths and weaknesses. The choice between them depends on the specific requirements of the recommendation system, the type of data available, and the user base. Hybrid approaches that combine collaborative filtering and content-based techniques are also common, aiming to leverage the strengths of both methods and mitigate their weaknesses.

In this mini-project, you'll be building both content based and collaborative filtering engines for the [MovieLens 25M dataset](https://grouplens.org/datasets/movielens/25m/). The MovieLens 25M dataset is one of the most widely used and popular datasets for building and evaluating recommendation systems. It is provided by the GroupLens Research project, which collects and studies datasets related to movie ratings and recommendations. The MovieLens 25M dataset contains movie ratings and other related information contributed by users of the MovieLens website.

**Dataset Details:**
- **Size:** The dataset contains approximately 25 million movie ratings.
- **Users:** It includes ratings from over 162,000 users.
- **Movies:** The dataset consists of ratings for more than 62,000 movies.
- **Ratings:** The ratings are provided on a scale of 1 to 5, where 1 is the lowest rating and 5 is the highest.
- **Timestamps:** Each rating is associated with a timestamp, indicating when the rating was given.

**Data Files:**
The dataset is usually split into three CSV files:

1. **movies.csv:** Contains information about movies, including the movie ID, title, genres, and release year.
   - Columns: movieId, title, genres

2. **ratings.csv:** Contains movie ratings provided by users, including the user ID, movie ID, rating, and timestamp.
   - Columns: userId, movieId, rating, timestamp

3. **tags.csv:** Contains user-generated tags for movies, including the user ID, movie ID, tag, and timestamp.
   - Columns: userId, movieId, tag, timestamp

First, import all the libraries you'll need.

In [2]:
import zipfile
import numpy as np
import pandas as pd
from urllib.request import urlretrieve
from sklearn.metrics.pairwise import cosine_similarity

Next, download the relevant components of the MoveLens dataset. Note, these instructions are roughly based on the colab [here](https://colab.research.google.com/github/google/eng-edu/blob/main/ml/recommendation-systems/recommendation-systems.ipynb?utm_source=ss-recommendation-systems&utm_campaign=colab-external&utm_medium=referral&utm_content=recommendation-systems#scrollTo=O3bcgduFo4s6).

In [3]:
print("Downloading movielens data...")

urlretrieve('http://files.grouplens.org/datasets/movielens/ml-100k.zip', 'movielens.zip')
zip_ref = zipfile.ZipFile('movielens.zip', 'r')
zip_ref.extractall()
print("Done. Dataset contains:")
print(zip_ref.read('ml-100k/u.info'))

ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv(
    'ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')

# The movies file contains a binary feature for each genre.
genre_cols = [
    "genre_unknown", "Action", "Adventure", "Animation", "Children", "Comedy",
    "Crime", "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror",
    "Musical", "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"
]
movies_cols = [
    'movie_id', 'title', 'release_date', "video_release_date", "imdb_url"
] + genre_cols
movies = pd.read_csv(
    'ml-100k/u.item', sep='|', names=movies_cols, encoding='latin-1')

Done. Dataset contains:
b'943 users\n1682 items\n100000 ratings\n'


Before doing any kind of machine learning, it's always good to familiarize yourself with the datasets you'lll be working with.

Here are your tasks:

1. Spend some time familiarizing yourself with both the `movies` and `ratings` dataframes. How many unique user ids are present? How many unique movies are there?
2. Create a new dataframe that merges the `movies` and `ratings` tables on 'movie_id'. Only keep the 'user_id', 'title', 'rating' fields in this new dataframe.

In [4]:
# Spend some time familiarizing yourself with both the movies and ratings
# dataframes. How many unique user ids are present? How many unique movies
# are there?
print(ratings.head())
print(ratings['user_id'].value_counts())
print(movies.head())
print(movies['movie_id'].value_counts())

   user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596
405    737
655    685
13     636
450    540
276    518
      ... 
441     20
36      20
812     20
895     20
93      20
Name: user_id, Length: 943, dtype: int64
   movie_id              title release_date  video_release_date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            imdb_url  genre_unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...              0       0   
1  http

It appears that there are 943 unique users in the ratings dataframe and 1682 unique movies in the movies dataframe.

In [5]:
# Merge movies and ratings dataframes
merged_df = pd.merge(ratings, movies, on='movie_id')[['user_id', 'title', 'rating']]

As mentioned in the introduction, content-Based Filtering is a recommendation engine approach that focuses on the attributes or features of items (products, movies, music, articles, etc.) and leverages these features to make personalized recommendations. The underlying idea is to match the characteristics of items with the preferences of users to suggest items that align with their interests. Content-based filtering is particularly useful when explicit user-item interactions (e.g., ratings or purchases) are sparse or unavailable.

**Key Steps in Content-Based Filtering:**

1. **Feature Extraction:**
   - For each item, relevant features are extracted. These features are typically descriptive attributes that can be represented numerically, such as genre, director, actors, author, publication date, and keywords.
   - In the case of text-based items, natural language processing techniques may be used to extract features like TF-IDF (Term Frequency-Inverse Document Frequency) scores.

2. **User Profile Creation:**
   - A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.
   - For example, if a user has watched several action movies, the action genre feature would receive a higher weight in their profile.

3. **Similarity Calculation:**
   - The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity, Euclidean distance, or Pearson correlation.
   - Cosine similarity is commonly used as it measures the cosine of the angle between two vectors, which represents their similarity.

4. **Recommendation:**
   - Items that are most similar to the user profile are recommended to the user. These are items whose features have the highest similarity scores with the user profile.
   - The recommended items are presented as a list sorted by their similarity scores.

**Advantages of Content-Based Filtering:**
1. **No Cold-Start Problem:** Content-based filtering can make recommendations even for new users with no historical interactions because it relies on item features rather than user history.

2. **User Independence:** The recommendations are based solely on the features of items and do not require knowledge of other users' preferences or behavior.

3. **Transparency:** Content-based recommendations are interpretable, as they depend on the features of items, making it easier for users to understand why specific items are recommended.

4. **Serendipity:** Content-based filtering can recommend items with characteristics not seen before by the user, leading to serendipitous discoveries.

5. **Diversity in Recommendations:** The method can offer diverse recommendations since it suggests items with different feature combinations.

**Limitations of Content-Based Filtering:**
1. **Limited Discovery:** Content-based filtering may struggle to recommend items outside the scope of users' historical interactions or interests.

2. **Over-Specialization:** Users may receive recommendations that are too similar to their previous choices, leading to a lack of exposure to new item categories.

3. **Dependency on Feature Quality:** The quality and relevance of item features significantly influence the quality of recommendations.

4. **Limited for Cold Items:** Content-based filtering can struggle to recommend new items with limited feature information.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return content-based recommendations for this user. Here are steps you can take:

  A. Get the user's rated movies

  B. Create a TF-IDF matrix using movie genres. Note, this can be extracted from the `movies` dataframe.

  C. Compute the cosine similarity between movie genres. Use the [cosine_similarity](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html) function.

  D. Get the indices of similar movies to those rated by the user based on cosine similarity. Keep only the top 5.

  E. Remove duplicates and movies already rated by the user.

In [17]:
'''before our content based recommendation function will work, we need
to make a new column in the movies df that has a text list of genres'''

'''Create a new column 'genres' containing a list of genre column names with value 1, do not include
first 5 columns which are not genre related'''
genre_columns = movies.columns[5:]
movie_genre_info = movies[genre_columns]
movies['genres'] = movie_genre_info.apply(lambda row: str(movie_genre_info.columns[row == 1].tolist()), axis=1)
print(movies['genres'].head())

0    ['Animation', 'Children', 'Comedy']
1    ['Action', 'Adventure', 'Thriller']
2                           ['Thriller']
3          ['Action', 'Comedy', 'Drama']
4         ['Crime', 'Drama', 'Thriller']
Name: genres, dtype: object


In [47]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Content-Based Filtering using Movie Genres
def content_based_recommendation(user_id, df):
  # Get the user's rated movies
  rated_movies = df[merged_df['user_id'] == user_id]['title'].tolist()
  # Create a TF-IDF matrix using movie genres
  tfidf_vectorizer = TfidfVectorizer(stop_words='english')

  tfidf_matrix = tfidf_vectorizer.fit_transform(movies['genres'].fillna(''))
  # Compute the cosine similarity between movie genres
  cosine_similarities = cosine_similarity(tfidf_matrix, tfidf_matrix)
  # Get the indices of the similar movies based on cosine similarity
  similar_movies_indices = []
  for movie_title in rated_movies:
      movie_index = movies.index[movies['title'] == movie_title].tolist()[0]
      similar_movies_indices += cosine_similarities[movie_index].argsort()[-6:-1].tolist()
      #print(similar_movies_indices)
  # Remove duplicates and movies already rated by the user
  recommendations = set(similar_movies_indices) - set(ratings[ratings['user_id'] == user_id].index.tolist())
  recommendations = list(recommendations)[:5]
  return recommendations

The key idea behind collaborative filtering is that users who have agreed in the past will likely agree in the future. Instead of relying on item attributes or user profiles, collaborative filtering identifies patterns of user behavior and item preferences from the interactions present in the data.

**Types of Collaborative Filtering:**
There are two main types of collaborative filtering:

**Collaborative Filtering Process:**
The collaborative filtering process typically involves the following steps:

1. **Data Collection:**
   - Gather data on user-item interactions, such as movie ratings, product purchases, or article clicks.

2. **User-Item Matrix:**
   - Organize the data into a user-item matrix, where rows represent users, columns represent items, and the entries contain the users' interactions (e.g., ratings).

3. **Similarity Calculation:**
   - Calculate the similarity between users or items using similarity metrics such as cosine similarity, Pearson correlation, or Jaccard similarity.
   - For user-based collaborative filtering, user similarities are calculated, and for item-based collaborative filtering, item similarities are calculated.

4. **Neighborhood Selection:**
   - For each user or item, select the most similar users or items as the neighborhood.
   - The size of the neighborhood (the number of similar users or items to consider) is an important parameter to control the system's behavior.

5. **Prediction Generation:**
   - Predict the ratings for items that the target user has not yet interacted with by combining the ratings of neighboring users or items.

6. **Recommendation Generation:**
   - Recommend items with the highest predicted ratings to the target user.

**Advantages of Collaborative Filtering using User-Item Interactions:**
- Collaborative filtering is based solely on user interactions and does not require knowledge of item attributes, making it useful for cases where item data is sparse or unavailable.
- It can provide serendipitous recommendations, suggesting items that users may not have discovered on their own.
- Collaborative filtering can be applied in various domains, including e-commerce, music, movie, and content recommendations.

**Limitations of Collaborative Filtering:**
- The cold-start problem: Collaborative filtering struggles to recommend to new users or items with no or limited interaction history.
- It may suffer from sparsity when data is limited or when users have only interacted with a small subset of items.
- Scalability issues can arise with large datasets and an increasing number of users or items.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return collaborative filtering recommendations for this user based on a user-item interaction matrix. Here are steps you can take:

  A. Create the user-item matrix using Pandas' [pivot_table](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html).

  B. Fill missing values with zeros in this matrix.

  C. Calculate user-user similarity matrix using cosine similarity.

  D. Get the array of similarity scores of the target user with all other users from the similarity matrix.

  E. Extract, say the the top 5 most similar users (excluding the target user).

  F. Generate movie recommendations based on the most similar users.

  G. Remove duplicate movies recommendations.

In [73]:
# Collaborative Filtering using User-Item Interactions
def collaborative_filtering_recommendation(user_id, df):
  # Create the user-item matrix
  user_item_matrix = pd.pivot_table(df, values='rating', index='user_id', columns='title', fill_value=0)
  # Fill missing values with 0 (indicating no rating)
  user_item_matrix_filled = user_item_matrix.fillna(0)
  # Calculate user-user similarity matrix using cosine similarity
  user_similarity_matrix = cosine_similarity(user_item_matrix_filled)
  # Get the similarity scores of the target user with all other users
  target_user_similarity_scores = user_similarity_matrix[user_id - 1]
  # Find the top N most similar users (excluding the target user)
  similar_users_indices = target_user_similarity_scores.argsort()[:-6:-1]
  # Generate movie recommendations based on the most similar users
  recommendations = []
  # what we will do is add the top rated movie from each user to the list of recommendations
  for user_index in similar_users_indices:
      rated_movies_by_similar_user = user_item_matrix_filled.iloc[user_index]
      top_movies = rated_movies_by_similar_user[rated_movies_by_similar_user == rated_movies_by_similar_user.max()].index.tolist()
      recommendations.extend(top_movies)
  from collections import Counter
  # Use Counter to get top 5 most frequently occurring in list recommendations
  counter = Counter(recommendations)
  # Get the top 5 most common values
  top_values = [value for value, count in counter.most_common(5)]


  return top_values

Now, test your recommendations engines! Select a few user ids and generate recommendations using both functions you've written. Are the recommendations similar? Do the recommendations make sense?

In [57]:
# Test the recommendation engines

# first look at what test user user_id = 5 likes
filtered_df = merged_df[merged_df['user_id'] == 5][merged_df['rating'] == 5]
for title in filtered_df['title'].tolist():
  #print(title)
  print(movies.columns)
  print(movies.loc[movies['title'] == title, ['Action', 'Children', 'Drama', 'Comedy']])

Index(['movie_id', 'title', 'release_date', 'video_release_date', 'imdb_url',
       'genre_unknown', 'Action', 'Adventure', 'Animation', 'Children',
       'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir',
       'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War',
       'Western', 'genres'],
      dtype='object')
     Action  Children  Drama  Comedy
256       1         0      0       1
Index(['movie_id', 'title', 'release_date', 'video_release_date', 'imdb_url',
       'genre_unknown', 'Action', 'Adventure', 'Animation', 'Children',
       'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir',
       'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War',
       'Western', 'genres'],
      dtype='object')
    Action  Children  Drama  Comedy
99       0         0      1       0
Index(['movie_id', 'title', 'release_date', 'video_release_date', 'imdb_url',
       'genre_unknown', 'Action', 'Adventure', 'Animation', 'Childre

<ipython-input-57-f8f24cc7a36d>:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  filtered_df = merged_df[merged_df['user_id'] == 5][merged_df['rating'] == 5]


It appears the user with user_id == 5 is fond of the comedy genre. So we might expect to see some comedies get recommended from our engine choices, especially with content-based recommendation.

In [59]:
user_id_to_recommend_content = 5
content_based_recommendations = content_based_recommendation(user_id_to_recommend_content, merged_df)

print(f"Content-Based Recommendations for User {user_id_to_recommend_content}:")
print(content_based_recommendations)

Content-Based Recommendations for User 5:
[1, 514, 1026, 1539, 1029]


These are the movie IDs of the top 5 recommended movies for user_id = 100. But which movies are these?

In [60]:
# Filter rows in column 'title' where the value in column 'movie_id' is in list content_based_recommendations
filtered_values = movies[movies['movie_id'].isin(content_based_recommendations)]['title'].tolist()
print(filtered_values)

['Toy Story (1995)', 'Annie Hall (1977)', 'Lay of the Land, The (1997)', 'Jury Duty (1995)', 'Being Human (1993)']


Are any of these comedies?

In [61]:
# Filter rows in column 'genres' where the value in column 'movie_id' is in list content_based_recommendations
filtered_values = movies[movies['movie_id'].isin(content_based_recommendations)]['genres'].tolist()
print(filtered_values)

["['Animation', 'Children', 'Comedy']", "['Comedy', 'Romance']", "['Comedy', 'Drama']", "['Comedy']", "['Drama']"]


We do see that 4 out of the top 5 recommended films are comedies.

Now let's see what results we get out of the collaborative filtering recommendation.

In [80]:
user_id_to_recommend_content = 5
collaborative_filtering_recommendations = collaborative_filtering_recommendation(user_id_to_recommend_content, merged_df)

print(f"Collaborative Filtering Recommendations for User {user_id_to_recommend_content}:")
print(collaborative_filtering_recommendations)

Collaborative Filtering Recommendations for User 5:
['Empire Strikes Back, The (1980)', 'Return of the Jedi (1983)', 'Wrong Trousers, The (1993)', 'Monty Python and the Holy Grail (1974)', 'Princess Bride, The (1987)']


In [79]:
# Filter rows in column 'genres' where the value in column 'movie_id' is in list content_based_recommendations
filtered_values = movies[movies['title'].isin(collaborative_filtering_recommendations)]['genres'].tolist()
print(filtered_values)

["['Comedy']", "['Animation', 'Comedy']", "['Action', 'Adventure', 'Drama', 'Romance', 'Sci-Fi', 'War']", "['Action', 'Adventure', 'Comedy', 'Romance']", "['Action', 'Adventure', 'Romance', 'Sci-Fi', 'War']"]


We see that the recommended movie set consists of comedies, action, dramas, sci-fi, and even romance. This recommendation is a little less explainable but is indeed more surendipitous than the content-based apporoach. Personally, I think the collaborative filtering based approach is more interesting but there is value in using both methods. I could see an optimized system using both in a sort of hybrid model.